contracts/zero-ex/contracts/src/features/multiplex/BatchMultiplexFeature.sol

```diff
@@ -0,0 +1,211 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+  Copyright 2024 ZeroEx Intl., Rigo Intl.
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+pragma solidity ^0.6.5;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-utils/contracts/src/v06/LibSafeMathV06.sol";
+import "../../examples/BatchMultiplexValidator.sol";
+import "../../fixins/FixinCommon.sol";
+import "../../fixins/FixinReentrancyGuard.sol";
+import "../../migrations/LibMigrate.sol";
+import "../interfaces/IFeature.sol";
+import "../interfaces/IBatchMultiplexFeature.sol";
+
+/// @dev This feature enables batch transactions by re-routing the single swaps to the exchange proxy.
+contract BatchMultiplexFeature is IFeature, IBatchMultiplexFeature, FixinCommon, FixinReentrancyGuard {
+    bytes4 private constant _VALIDATE_SELECTOR = BatchMultiplexValidator.validate.selector;
+
+    // TODO: remove as this is mock for validator setup and tests
+    address public immutable _validator;
+
+    /// @inheritdoc IFeature
+    string public constant override FEATURE_NAME = "BatchMultiplexFeature";
+    /// @inheritdoc IFeature
+    uint256 public immutable override FEATURE_VERSION = _encodeVersion(1, 0, 0);
+
+    /// @dev Refunds up to `msg.value` leftover ETH at the end of the call.
+    modifier refundsAttachedEth() {
+        _;
+        uint256 remainingBalance = LibSafeMathV06.min256(msg.value, address(this).balance);
+        if (remainingBalance > 0) {
+            msg.sender.transfer(remainingBalance);
+        }
+    }
+
+    /// @dev Ensures that the ETH balance of `this` does not go below the
+    ///      initial ETH balance before the call (excluding ETH attached to the call).
+    modifier doesNotReduceEthBalance() {
+        uint256 initialBalance = address(this).balance - msg.value;
```

Comment:
since the balance could be 0, the value would be negative and so the `initialBalance` type needs to be signed I think

Reply:
technically that is not correct, as the proxy's ETH balance is increased by the msg.value and `initialBalance` is defined before calls execution. The proxy balance will always be the balance before the batchMultiplex call, plus the msg.value of the batchMultiplex call, i.e. `initialBalance` will always be at least as big as `msg.value`. Proxy ETH balance would be 0 only if msg.value is 0, which does not result in underflow. If the calc did underflow, the post-execution assertion `require(initialBalance <= address(this).balance, "Batch_M_Feat/ETH_LEAK");` would revert as initialBalance would be greater than address(this).balance

Reply:
got It ty!